### PR TITLE
net: l2: wifi_mgmt: remove scan kconfig defaults

### PR DESCRIPTION
**Pull request by JordanYates**
Remove the wifi_mgmt interface overriding default values with values from kconfig.

These defaults were only applied when a `params` struct was provided by the application. This is the case when the application is explicitly setting the options it wants, why is the mgmt API changing these. When `params` is NULL and thus modem defaults are requested, these defaults aren't applied. This is the opposite behaviour from what seems reasonable.

In addition, these options are:
 * Undocumented
 * Using non-trivial string parsing functions (strtok)
 * Adding complexity to the API implementation by forcing support for ROM versions of command line arguments.

**Additional comments outside the commit message**

This overriding was started in #59204 and continued in #60686.
Neither PR justified overriding the application request from what I can see.
None of the reviews questioned why it was being added at all as far as I can see.

From my perspective, I see no reason why the management layer ever needs to mess with the scan request. The application is asking for it, if it wants to provide some defaults, do that in the request, not in some kconfig option (which is only applied some of the time).

Apart from being incorrect in my opinion, supporting these defaults is also making cleaning up the massive RAM usage of the new struct problematic because the "defaults" being applied to `struct wifi_params` needs to be created from ROM strings.

### CODE REVIEW

**Comment by JordanYates** [comment]
Pinging @ajayparida and @sachinthegreen as the authors of the relevant PR's

**Comment by krish2718** [comment]
> Neither PR justified overriding the application request from what I can see.
None of the reviews questioned why it was being added at all as far as I can see.

From my perspective, I see no reason why the management layer ever needs to mess with the scan request. The application is asking for it, if it wants to provide some defaults, do that in the request, not in some kconfig option (which is only applied some of the time).

IIUC, it's to provide both runtime and compile time overrides, with compile time overrides being highest priority. E.g., a customer  knows that he has deployed only 2.4GHz Wi-Fi network, so, he can release an image that supports 2.4GHz scan only independent of if the application (written by end customers) asks for 2.4/5GHz bands scan.

**Comment by JordanYates** [comment]
> IIUC, it's to provide both runtime and compile time overrides, with compile time overrides being highest priority. E.g., a customer knows that he has deployed only 2.4GHz Wi-Fi network, so, he can release an image that supports 2.4GHz scan only independent of if the application (written by end customers) asks for 2.4/5GHz bands scan.

Except this is not documented anywhere, and is not actually what it does.
* `net_mgmt(NET_REQUEST_WIFI_SCAN, iface, NULL, 0)` does not apply any of these options.
* Compile time overrides are the lowest priority, not the highest.

What image is being released here? Who are these two customers? I am still not seeing the point.

**Comment by krish2718** [comment]
> > IIUC, it's to provide both runtime and compile time overrides, with compile time overrides being highest priority. E.g., a customer knows that he has deployed only 2.4GHz Wi-Fi network, so, he can release an image that supports 2.4GHz scan only independent of if the application (written by end customers) asks for 2.4/5GHz bands scan.
> 
> Except this is not documented anywhere, and is not actually what it does.

Yes, the behaviour of runtime vs compile time overrides isn't documented, we can improve that.
> 
> * `net_mgmt(NET_REQUEST_WIFI_SCAN, iface, NULL, 0)` does not apply any of these options.

That is a special case mainly to maintain backwards compatibility.

> * Compile time overrides are the lowest priority, not the highest.

It depends on the parameters and whether if the override makes sense. E.g., `CONFIG_WIFI_MGMT_FORCED_PASSIVE_SCAN` overrides the setting from the API. `CONFIG_WIFI_MGMT_SCAN_CHAN` run time is highest and compile time is used as a fallback.

> What image is being released here? Who are these two customers? I am still not seeing the point.

I was just trying to paint a picture, so, only gave you the overview, you need to imagine the gaps :). 

Scenario: Zephyr customer- A who is an OEM is working with multiple end-users/customers B,C,D.... For a given environment/site for customer B, Wi-Fi deployment is limited to 2.4GHz only, so, no point in scanning 5GHz which is the default (scan all supported bands), this can be achieved in two ways.

1. Modify the application source (Developed by A) specific to B to request 2.4GHz only scan, but this poses maintenance issues. And if the number of applications are more then this change has to be done for all of them.
2. Keep the source unchanged, generate multiple Zephyr images with configuration overlays for each customer in this case B with 2.4GHz channel list.

We aren't enforcing a specific way, just giving options both runtime and compile time, except for the missing docs on how overrides work (Welcome any PRs related to this), I don't see any issue here and having these overrides is quite handy.

**Comment by JordanYates** [comment]
> Scenario: Zephyr customer- A who is an OEM is working with multiple end-users/customers B,C,D.... For a given environment/site for customer B, Wi-Fi deployment is limited to 2.4GHz only, so, no point in scanning 5GHz which is the default (scan all supported bands), this can be achieved in two ways.
> 
> 1. Modify the application source (Developed by A) specific to B to request 2.4GHz only scan, but this poses maintenance issues. And if the number of applications are more then this change has to be done for all of them.
> 2. Keep the source unchanged, generate multiple Zephyr images with configuration overlays for each customer in this case B with 2.4GHz channel list.

Why is this behaviour not an application level kconfig which sets the scan parameters to the desired values?

I understand that the situation can be solved the way you have done it here, it just seems an anti-pattern compared to all other Zephyr code I have seen. I have not seen API's inserting themselves between the call and the driver, and I certainly haven't seen them do so by adding custom string parsing rules and formats. I don't care if someone wants to insert shell utility functions into their application, that's their decision, I just don't think they make sense in the core management code.

I can maybe see the argument for `CONFIG_WIFI_MGMT_FORCED_PASSIVE_SCAN` in the context of EMC certifications, but this kind of falls over when it isn't even supported by most drivers. The rest of the options flat out aren't relevant for any in-tree driver.

**Comment by JordanYates** [comment]
To give you my perspective, I am trying to clean up a gross oversight from the PR's where this functionality was added (#62735, which you have approved) but am unable to do so because of this functionality (which I happen to disagree with anyway).

It is not possible to convert a `"ssid1;ssid2"` in ROM into `char *`, as would be required to keep this functionality.

So instead I am needing to argue for why this should be removed, when no-one provided arguments or agreement as to why it should be included in the first place.

**Comment by krish2718** [comment]
> > Scenario: Zephyr customer- A who is an OEM is working with multiple end-users/customers B,C,D.... For a given environment/site for customer B, Wi-Fi deployment is limited to 2.4GHz only, so, no point in scanning 5GHz which is the default (scan all supported bands), this can be achieved in two ways.
> > 
> > 1. Modify the application source (Developed by A) specific to B to request 2.4GHz only scan, but this poses maintenance issues. And if the number of applications are more then this change has to be done for all of them.
> > 2. Keep the source unchanged, generate multiple Zephyr images with configuration overlays for each customer in this case B with 2.4GHz channel list.
> 
> Why is this behaviour not an application level kconfig which sets the scan parameters to the desired values?

Sure it can be done anywhere, the rationale for putting it in the stack is that this is a common Wi-Fi features, so, agnostic to applications/drivers. Applications can configure as per their needs, just that the compile time overrides seemed appropriate in the stack.

> I understand that the situation can be solved the way you have done it here, it just seems an anti-pattern compared to all other Zephyr code I have seen. I have not seen API's inserting themselves between the call and the driver, and I certainly haven't seen them do so by adding custom string parsing rules and formats. I don't care if someone wants to insert shell utility functions into their application, that's their decision, I just don't think they make sense in the core management code.

I don't know the specific things you are referring here, but AFAICS, Zephyr (at least networking) has this pattern, this isn't the same as the one being discussed here, but it does show API's doing compile time checks rather than simple relay. And for this case of Wi-Fi scan, there is a significant impact of time with each flag hence the reason for the complex input through shell/Kconfig/API.

```
func api_always_available:
  #ifdef CONFIG_FEAT_A
        do():
   #else
       not supported
  #endif
 ```

> I can maybe see the argument for `CONFIG_WIFI_MGMT_FORCED_PASSIVE_SCAN` in the context of EMC certifications, but this kind of falls over when it isn't even supported by most drivers. The rest of the options flat out aren't relevant for any in-tree driver.

True, the in-tree drivers are quite basic and don't expose any options than, scan/connect/disconnect. This entire feature is to extend the Wi-Fi stack for the upcoming in-tree drivers that aren't fully offloaded and utilise these (if and when that happens I don't know).

TL;DR, I still think these compile time overrides are useful, e just need to document them. And your PRs address the extra memory overhead associated with that. If that is still not acceptable, we could just put all of these under `WIFI_ADVANCED_SCAN_COMPILE_FLAGS` feature flag and possibly a separate `_overrides` function to make core readable. WDYT?

**Comment by krish2718** [comment]
> To give you my perspective, I am trying to clean up a gross oversight from the PR's where this functionality was added (#62735, which you have approved) but am unable to do so because of this functionality (which I happen to disagree with anyway).

It's not an oversight, its a deliberate decision to use those Kconfig overrides, you might disagree with the rationale but that is a different issue.

> It is not possible to convert a `"ssid1;ssid2"` in ROM into `char *`, as would be required to keep this functionality.

Are you referring to the fact that it needs a `strdup/strcpy` of `CONFIG_**` strings? 
 
> So instead I am needing to argue for why this should be removed, when no-one provided arguments or agreement as to why it should be included in the first place.

I did provide the arguments (at least I thought I did), just that they are not aligned with your view. So, lets see what others think, there is no point in debating from same perspectives, it will lead no where.

**Comment by JordanYates** [comment]
> It's not an oversight, its a deliberate decision to use those Kconfig overrides, you might disagree with the rationale but that is a different issue.

The gross oversight was not the addition of this functionality, which while I disagree with I am happy to defer to consensus. The oversight was the bloat of `struct wifi_params` from 1 byte to 1.4kB, parts of which cannot be resolved because of this functionality.

> Are you referring to the fact that it needs a strdup/strcpy of CONFIG_** strings?

My preferred API for the SSIDs is `char *ssids[N]`, because the SSIDs are being cached by the driver, so why bloat the struct by another 64 bytes. `char *` implies that each SSID is NULL terminated, which cannot be obtained directly from `CONFIG_WIFI_MGMT_SCAN_SSID_FILT` as that lives in ROM and `;` is obviously not a `\0`.

> I did provide the arguments (at least I thought I did), just that they are not aligned with your view.

I agree that you have provided them here, but they were not presented in the PR's where they were introduced (and you weren't the one who introduced them). My annoyance is that despite not being discussed, it is now on me to justify removing them, instead of on the authors for why they should be introduced in the first place.

I don't think that the concept overall is a bad idea, just that the implementation is clunky and doesn't match Zephyr standards. I would be 100% on board with some `wifi_scan_with_overrides_function` that does this, which I can then ignore.

But if @jukkar comes along and says this was all well considered in the original PR's and it stays, then that's his decision as the maintainer.

**Review by jukkar — APPROVED** [review]
As the original PRs did not had clear use cases explained and the changes are a bit controversial as seen in the discussion, I am inclined to accept this one. The changes could be re-introduced with proper use cases and reviews in the future.

**Comment by krish2718** [comment]
> My annoyance is that despite not being discussed, it is now on me to justify removing them, instead of on the authors for why they should be introduced in the first place.

I don't quite agree, isn't that the whole point of open-source where few things slip in which aren't right and then the community fixes them with a healthy discussion? (I call this discussion rather asking for justification).

Anyways, as @jukkar says let's merge this PR and we can re-introduce these changes with documented use cases  and behaviours.

**Review by krish2718 — APPROVED** [review]
